File: web/src/lib/api/invites.ts
import { z } from 'zod';

import { api } from '@/lib/api/client';
import { personColorSchema } from '@/lib/api/schemas';
import { placeSnapshotSchema } from '@/lib/api/dates';

/** Приглашение: отправка автором и публичный ответ. */

export const sendResultSchema = z.object({
  token: z.string(),
  url: z.string(),
  expires_at: z.string(),
});

export const invitePublicSchema = z.object({
  author_name: z.string(),
  author_color: personColorSchema,
  scheduled_at: z.string(),
  is_all_day: z.boolean(),
  note: z.string().nullable(),
  place: placeSnapshotSchema,
  answered: z.boolean(),
  accepted: z.boolean(),
});

export type InvitePublic = z.infer<typeof invitePublicSchema>;

export function sendInvite(dateId: string) {
  return api(`/dates/${dateId}/send`, { method: 'POST', schema: sendResultSchema });
}

/**
 * Публичные вызовы. `skipRefresh` обязателен: экран открывают без сессии,
 * и попытка обновить её увела бы человека в ошибку вместо приглашения.
 */
export function fetchInvite(token: string) {
  return api(`/invites/${token}`, { schema: invitePublicSchema, skipRefresh: true });
}

export function respondToInvite(token: string, accepted: boolean, evadeCount: number) {
  return api(`/invites/${token}/respond`, {
    method: 'POST',
    body: { accepted, evade_count: evadeCount },
    schema: invitePublicSchema,
    skipRefresh: true,
  });
}

File: api/app/schemas/invites.py
"""Схемы приглашений (ТЗ 11)."""

from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.dates import PlaceSnapshot


class SendResult(BaseModel):
    """Ответ на отправку приглашения — ссылка для share-sheet."""

    token: str
    url: str
    expires_at: datetime


class InvitePublic(BaseModel):
    """Данные свидания для публичного экрана.

    ID пользователей сюда **не попадают** (ТЗ 16): по ссылке видно только
    отображаемое имя и цвет автора. Знание токена даёт доступ к свиданию,
    но не к внутренним идентификаторам.
    """

    author_name: str
    author_color: str
    scheduled_at: datetime
    is_all_day: bool
    note: str | None
    place: PlaceSnapshot
    # Уже отвеченное приглашение открывается повторно и показывает исход.
    answered: bool
    accepted: bool


class InviteResponse(BaseModel):
    accepted: bool
    # Сколько раз убежала кнопка «Нет» — для истории, не для логики.
    evade_count: int = Field(default=0, ge=0, le=100)
